aip/0180.md

```diff
@@ -0,0 +1,126 @@
+The guidance here **should** be treated as indicative, rather than as a
+comprehensive list of every possible change.
+
+There are three distinct types of compatibility to consider:
+
+1. Source compatibility: Code written against a previous version **must**
+   compile against a newer version, and successfully run with a newer version
+   of the client library.
+2. Wire compatibility: Code written against a previous version **must** be able
+   to communicate correctly with a newer server. In other words, not only are
+   inputs and outputs compatible, but the serialization and deserialization
+   expectations continue to match.
+3. Semantic compatibility: Code written against a previous version **must**
+   continue to receive the intended and expected results.
+
+### Adding components
+
+In general, new components (interfaces, methods, messages, fields, enums, or
+enum values) **may** be added to existing APIs in the same major version.
+
+However, keep the following guidelines in mind when doing this:
+
+- Code written against the previous surface (and thus is unaware of the new
+  components) **must** continue to be treated the same way as before.
+  - New required fields **must not** be added to existing request messages or
+    resources.
+  - Any field being populated by clients **must** have a default behavior
+    matching the behavior before the field was introduced.
+  - Any field previously populated by the server **must** continue to be
+    populated, even if it introduces redundancy.
+- Enums **may** be freely added to request messages. For response messages and
+  resources, be aware that it is possible that user code does not handle new
+  values gracefully. In this situation, enums which are expected to receive new
+  values **should** document this, and while enum values still **may** be added
+  in this situation, appropriate caution **should** be used.
+
+**Note:** It is possible when adding a component closely related to an existing
+component (for example, `string foo_value` when `string foo` already exists) to
+enter a situation where generated code will conflict. Service owners **should**
+be aware of subtleties in the tooling they or their users are likely to use
+(and tool authors **should** endeavor to avoid such subtleties if possible).
+
+### Removing or renaming components
+
+Existing components (interfaces, methods, messages, fields, enums, or enum
+values) **must not** be removed from existing APIs in the same major version.
+Removing a component is a backwards incompatible change.
+
+**Important:** Renaming a component is semantically equivalent to "remove and
+add". In cases where these sorts of changes are desirable, a service **may**
+add the new component, but **must not** remove the existing one.
+
+### Changing the type of fields
+
+Existing fields and messages **must not** have their type changed, even if the
+new type is wire-compatible, because type changes alter generated code in a
```

Comment:
Is it worth noting that "wire compatible" means different things in JSON and proto? Are we writing AIPs in a way that assumes those are the two-and-only-two ways of calling an API?

Reply:
I initially thought about doing that and decided against it. I could be convinced otherwise.

Reply:
Let's chat in person about the scope of AIPs and what we mean by "wire compatible". It will affect more than just this one, probably.

Reply:
Discussed in person, and added a note.